### DISCUSSION TITLE
Upgrade sub with wiring

### DISCUSSION THREAD

**Discussion by duarm** [FakeFishGames/Barotrauma#12324] [Ideas and suggestions]
Instead of having an arbitrary UI where you buy upgrades to your sub, make the NPC instead sell "Upgrade Gates", things like SPEED+ Upgrade, EFFICIENCY+ Upgrade and others. The electrician then places that in the sub similar to the AND gate and the OR gate, and wires that to the Engine for example. Each machine would have x additional IN slots for upgrades, or an exclusive UPGRADE slot in the circuit.

This would:
- increase immersion
- makes these two mechanics more tightly designed, wiring is used in multiple systems
- might make the upgrade path less linear, as opposed to the "make me better at the game" style of upgrade we currently have. By that I mean, you need to consider what to place on each slot, as you can't have everything. Each upgrade might also have negative buffs, e.g. more speed = more energy consumption. If you don't think it's worth, you can always disconnect the gate, unlike the current system.

**Comment by Arctodites** [comment]
Such a system would also open up all kinds of possibilities for more sophisticated calculations, because we'd presumably be able to feed the info about what upgrades are installed into our circuit designs to calculate things like maximum hull pressure depth, maximum output of power devices, etc.  Right now, it's only possible to account for rated maximum pump speeds (flow rates) or rated reactor power outputs based upon the initial design values of the installations.  With your proposed system, we'd be able to feed data about upgrades into our ships' computation circuits.

**Reply by duarm** [reply]
Sure, what I had in mind was to make each upgrade have a negative effect, so we would have "submarine builds" or something like that, since you can't just throw more upgrades into the thing to make it better, without considering the side effects. It would be stressful to do that currently, since upgrades are permanent.

**Comment by gerakis** [comment]
As someone who does a lot of wiring, this concept doesn't really make sense. The signal outputs are just a number. An "upgrade" to the number is total nonsense. You can already make the number higher though normal wiring, it simply wont help you. Like an engine cannot go past 100% thrust even if you send it 200.

**Reply by duarm** [reply]
It doesn't affect the wiring or the signal, it's just a more immersive way to upgrade the engine, it will affect the engine the same way the engine upgrade you currently buy at stations.

**Comment by neirenoir** [comment]
While maybe logic gates are not the way to go about it, I like the idea of having the mechanics and engineers installing the upgrades. Imagine you buy a "railgun kit" instead of just adding it to the sub by clicking a button, and the mechanics have to put on a diving suit and mount the railgun outside while the engineers wire everything up. It could be a campaign toggle, or simply an option to pay more/less at the dry dock to get it done for you.

I think drawing inspiration from Barocraftables could also be interesting. I haven't had this much fun being the engineer since I installed the ability to craft and move junction boxes. With vanilla mechanics it's usually hard to find a junction box in the sub that isn't already full and overworked to install a single relay component, but being able to craft them allows me to redo part of the grid to my liking. Being able to maybe buy more of them, as well as pumps or "enginelets" (aux. engines mechanics could install outside the sub) as an upgrade or recovered from enemy subs/wrecks instead of the current system could spice up campaign gameplay. After all, the sub editor is like half of the game: exposing some of its already existing mechanics in-game so the crew can truly feel they are working together to improve the efficiency of their sub.

**Reply by duarm** [reply]
I like the idea. My point is to just use very simple wiring to interact with mechanics like this, it's a missed oportunity otherwise, of course you can use have something which looks like an electric box for example, instead of a gate
